File: routa-core/src/main/kotlin/com/phodal/routa/core/provider/AcpAgentProvider.kt
package com.phodal.routa.core.provider

import com.agentclientprotocol.common.Event
import com.agentclientprotocol.model.SessionUpdate
import com.phodal.routa.core.acp.AcpProcessManager
import com.phodal.routa.core.acp.RoutaAcpClient
import com.phodal.routa.core.config.AcpAgentConfig
import com.phodal.routa.core.model.AgentRole
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.io.asSink
import kotlinx.io.asSource
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

/**
 * ACP-based agent provider with full lifecycle management.
 *
 * Compared to the legacy [com.phodal.routa.core.runner.AcpAgentRunner], this adds:
 * - **Health check**: monitors whether the ACP process is alive
 * - **Interrupt**: can kill a running agent process
 * - **Streaming**: delivers [StreamChunk] events in real-time during execution
 * - **Session tracking**: maintains per-agent connection state for cleanup
 *
 * Modeled after Intent's `ACPProvider` which manages child processes, handles
 * session recovery, and provides health monitoring.
 *
 * ## Process Lifecycle
 * ```
 * run()/runStreaming() → spawn process → ACP connect → create session → prompt → collect events → disconnect
 * interrupt()         → kill process
 * isHealthy()        → check process.isAlive
 * cleanup()          → terminate process, clear state
 * ```
 */
class AcpAgentProvider(
    private val agentKey: String,
    private val config: AcpAgentConfig,
    private val cwd: String,
) : AgentProvider {

    private val processManager = AcpProcessManager.getInstance()

    // Track active connections per agentId for health checks and cleanup
    private val activeClients = ConcurrentHashMap<String, ActiveAcpSession>()

    // Track last heartbeat per agentId
    private val lastHeartbeats = ConcurrentHashMap<String, Instant>()

    private data class ActiveAcpSession(
        val processKey: String,
        val client: RoutaAcpClient,
        val scope: CoroutineScope,
        val startedAt: Instant = Instant.now(),
    )

    // ── AgentRunner (backward compat) ────────────────────────────────

    override suspend fun run(role: AgentRole, agentId: String, prompt: String): String {
        return runStreaming(role, agentId, prompt) { /* discard chunks */ }
    }

    // ── AgentProvider: Streaming ─────────────────────────────────────

    override suspend fun runStreaming(
        role: AgentRole,
        agentId: String,
        prompt: String,
        onChunk: (StreamChunk) -> Unit,
    ): String {
        val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())
        val processKey = "$agentKey-$agentId"

        // Spawn (or reuse) the agent process
        val managed = processManager.getOrCreateProcess(
            agentKey = processKey,
            command = config.getCommandLine(),
            cwd = cwd,
            env = config.env,
        )

        // Connect via ACP protocol
        val client = RoutaAcpClient(
            coroutineScope = scope,
            input = managed.inputStream.asSource(),
            output = managed.outputStream.asSink(),
            cwd = cwd,
            agentName = agentKey,
        )

        // Track this session
        activeClients[agentId] = ActiveAcpSession(processKey, client, scope)
        updateHeartbeat(agentId)

        val resultBuilder = StringBuilder()

        // Wire up streaming callbacks
        client.onSessionUpdate = { update ->
            updateHeartbeat(agentId)
            when (update) {
                is SessionUpdate.AgentMessageChunk -> {
                    val text = RoutaAcpClient.extractText(update.content)
                    resultBuilder.append(text)
                    onChunk(StreamChunk.Text(text))
                }
                is SessionUpdate.ToolCallUpdate -> {
                    val toolName = update.title ?: "tool"
                    val statusStr = update.status.toString().lowercase()
                    val status = when {
                        statusStr.contains("start") -> ToolCallStatus.STARTED
                        statusStr.contains("complet") -> ToolCallStatus.COMPLETED
                        statusStr.contains("fail") || statusStr.contains("error") -> ToolCallStatus.FAILED
                        else -> ToolCallStatus.IN_PROGRESS
                    }
                    onChunk(StreamChunk.ToolCall(toolName, status))
                }
                else -> {
                    // Plan updates, mode changes, thought chunks, etc.
                    // Emit heartbeat so coordinator knows agent is alive
                    onChunk(StreamChunk.Heartbeat())
                }
            }
        }

        try {
            client.connect()
            onChunk(StreamChunk.Heartbeat()) // Connected

            // Send the task prompt
            client.prompt(prompt).collect { event ->
                updateHeartbeat(agentId)
                when (event) {
                    is Event.PromptResponseEvent -> {
                        val stopReason = event.response.stopReason
                        onChunk(StreamChunk.Completed(stopReason.toString()))
                    }
                    is Event.SessionUpdateEvent -> {
                        // Already handled by onSessionUpdate callback
                    }
                }
            }
        } catch (e: Exception) {
            onChunk(StreamChunk.Error(e.message ?: "ACP error", recoverable = true))
            throw e
        } finally {
            client.disconnect()
            processManager.terminateProcess(processKey)
            activeClients.remove(agentId)
            scope.coroutineContext[Job]?.cancel()
        }

        return resultBuilder.toString().ifEmpty {
            "[ACP agent completed without text output]"
        }
    }

    // ── AgentProvider: Health Check ──────────────────────────────────

    override fun isHealthy(agentId: String): Boolean {
        val session = activeClients[agentId] ?: return false
        // Check 1: Is the OS process alive?
        if (!processManager.isRunning(session.processKey)) return false
        // Check 2: Have we received a heartbeat recently? (5 min staleness threshold)
        val lastBeat = lastHeartbeats[agentId] ?: return false
        val staleThresholdMs = 5 * 60 * 1000L
        return Instant.now().toEpochMilli() - lastBeat.toEpochMilli() < staleThresholdMs
    }

    // ── AgentProvider: Interrupt ─────────────────────────────────────

    override suspend fun interrupt(agentId: String) {
        val session = activeClients.remove(agentId)
        if (session != null) {
            try {
                session.client.disconnect()
            } catch (_: Exception) { }
            processManager.terminateProcess(session.processKey)
            session.scope.coroutineContext[Job]?.cancel()
        }
        lastHeartbeats.remove(agentId)
    }

    // ── AgentProvider: Capabilities ──────────────────────────────────

    override fun capabilities(): ProviderCapabilities = ProviderCapabilities(
        name = "ACP ($agentKey)",
        supportsStreaming = true,
        supportsInterrupt = true,
        supportsHealthCheck = true,
        supportsFileEditing = true,
        supportsTerminal = true,
        supportsToolCalling = false, // ACP agents handle tools internally
        maxConcurrentAgents = 5,
        priority = 10,
    )

    // ── AgentProvider: Cleanup ───────────────────────────────────────

    override suspend fun cleanup(agentId: String) {
        interrupt(agentId) // Reuse interrupt for cleanup
    }

    override suspend fun shutdown() {
        val agents = activeClients.keys.toList()
        for (agentId in agents) {
            interrupt(agentId)
        }
    }

    // ── Internal ─────────────────────────────────────────────────────

    private fun updateHeartbeat(agentId: String) {
        lastHeartbeats[agentId] = Instant.now()
    }
}

File: routa-js/src/core/provider/acp-agent-provider.ts
/**
 * AcpAgentProvider - port of routa-core AcpAgentProvider.kt
 *
 * ACP-based agent provider for CRAFTER and GATE roles.
 * Connects to ACP agents like opencode, codex, etc.
 *
 * This implementation:
 * - Connects to existing AcpProcessManager
 * - Creates separate sessions for each agent
 * - Uses session/set_mode to switch between 'plan' and 'build' modes
 * - Handles streaming via session/update notifications
 */

import { AgentRole } from "../models/agent";
import {
  AgentProvider,
  ProviderCapabilities,
  StreamChunk,
} from "./agent-provider";
import { AcpProcessManager } from "../acp/acp-process-manager";
import { JsonRpcMessage } from "../acp/processer";

/**
 * Configuration for ACP agent
 */
export interface AcpAgentConfig {
  /** Preset ID (e.g., "opencode", "codex") */
  presetId: string;

  /** Working directory */
  cwd: string;

  /** Environment variables */
  env?: Record<string, string>;

  /** Extra command-line arguments */
  extraArgs?: string[];
}

/**
 * Managed session info
 */
interface ManagedSession {
  /** Our internal session ID */
  sessionId: string;
  /** ACP session ID from the agent */
  acpSessionId: string;
  /** Agent role */
  role: AgentRole;
  /** Current mode (plan or build) */
  mode: string;
  /** Accumulated output */
  output: string;
  /** Creation timestamp */
  createdAt: Date;
}

/**
 * ACP-based agent provider.
 *
 * Connects to the existing AcpProcessManager and creates sessions
 * with appropriate modes based on agent role:
 * - ROUTA (planner) → 'plan' mode (read-only, no file editing)
 * - CRAFTER (implementer) → 'build' mode (full tools)
 * - GATE (reviewer) → 'plan' mode (read-only verification)
 */
export class AcpAgentProvider implements AgentProvider {
  private config: AcpAgentConfig;
  private processManager: AcpProcessManager;
  private activeSessions = new Map<string, ManagedSession>();

  constructor(config: AcpAgentConfig, processManager?: AcpProcessManager) {
    this.config = config;
    this.processManager = processManager ?? new AcpProcessManager();
  }

  async run(role: AgentRole, agentId: string, prompt: string): Promise<string> {
    return this.runStreaming(role, agentId, prompt, () => {});
  }

  async runStreaming(
    role: AgentRole,
    agentId: string,
    prompt: string,
    onChunk: (chunk: StreamChunk) => void
  ): Promise<string> {
    try {
      // Determine the mode based on role
      const mode = this.getModeForRole(role);

      // Accumulated output
      let output = "";

      // Create notification handler for streaming
      const onNotification = (msg: JsonRpcMessage) => {
        if (msg.method === "session/update") {
          const params = msg.params as any;
          const update = params.update || params;
          const sessionUpdate = update.sessionUpdate;

          // Handle different update types (matching Kotlin's processSessionUpdateWithEvents)
          if (sessionUpdate === "agent_message_chunk") {
            // Extract text from content block
            const content = update.content;
            let text = "";
            if (content && content.type === "text") {
              text = content.text || "";
            }
            output += text;
            onChunk({ type: "text", content: text });
          } else if (sessionUpdate === "agent_thought_chunk") {
            // Extract thought text from content block
            const content = update.content;
            let thought = "";
            if (content && content.type === "text") {
              thought = content.text || "";
            }
            // Send as thinking chunk (not accumulated in output)
            onChunk({ type: "thinking", content: thought });
          } else if (sessionUpdate === "tool_call_update") {
            // Handle tool call updates
            const toolName = update.title || "tool";
            const rawInput = update.rawInput ? JSON.stringify(update.rawInput) : "";
            const rawOutput = update.rawOutput ? JSON.stringify(update.rawOutput) : "";

            // Send tool call start
            if (update.status?.toString().toLowerCase().includes("start")) {
              onChunk({ type: "tool_call", name: toolName, args: {} });
            }
            // Send tool result when completed
            else if (update.status?.toString().toLowerCase().includes("complet")) {
              onChunk({ type: "tool_result", name: toolName, result: rawOutput || "completed" });
            }
          } else if (params.type === "error") {
            onChunk({ type: "error", message: params.error || "Unknown error" });
          }
        }
      };

      // Create session with ACP process manager
      const acpSessionId = await this.processManager.createSession(
        agentId,
        this.config.cwd,
        onNotification,
        this.config.presetId,
        this.config.extraArgs,
        this.config.env
      );

      // Get the managed process
      const managed = this.processManager.getManagedProcess(agentId);
      if (!managed) {
        throw new Error(`Failed to get managed process for agent ${agentId}`);
      }

      // Set the appropriate mode using session/set_mode
      console.log(`[AcpAgentProvider] Setting mode to '${mode}' for ${role} (agent: ${agentId})`);
      await managed.process.sendRequest("session/set_mode", {
        sessionId: acpSessionId,
        modeId: mode,
      });

      // Store session info
      this.activeSessions.set(agentId, {
        sessionId: agentId,
        acpSessionId,
        role,
        mode,
        output: "",
        createdAt: new Date(),
      });

      // Send the prompt
      console.log(`[AcpAgentProvider] Sending prompt to ${role} (agent: ${agentId})`);
      const result = await managed.process.prompt(acpSessionId, prompt);

      // Update session output
      const session = this.activeSessions.get(agentId);
      if (session) {
        session.output = output;
      }

      console.log(`[AcpAgentProvider] ${role} completed with stopReason: ${result.stopReason}`);
      return output || result.stopReason;
    } catch (error) {
      console.error(`[AcpAgentProvider] Error running ${role}:`, error);
      throw error;
    }
  }

  isHealthy(agentId: string): boolean {
    const managed = this.processManager.getManagedProcess(agentId);
    return managed?.process.alive ?? false;
  }

  async interrupt(agentId: string): Promise<void> {
    const session = this.activeSessions.get(agentId);
    if (!session) {
      return;
    }

    const managed = this.processManager.getManagedProcess(agentId);
    if (managed) {
      await managed.process.cancel(session.acpSessionId);
    }
  }

  capabilities(): ProviderCapabilities {
    return {
      name: `ACP (${this.config.presetId})`,
      supportsStreaming: true,
      supportsInterrupt: true,
      supportsHealthCheck: true,
      supportsFileEditing: true,
      supportsTerminal: true,
      supportsToolCalling: false, // ACP agents handle tools internally
      maxConcurrentAgents: 5,
      priority: 10, // Higher priority for implementation tasks
    };
  }

  async cleanup(agentId: string): Promise<void> {
    await this.interrupt(agentId);
    this.processManager.killSession(agentId);
    this.activeSessions.delete(agentId);
  }

  async shutdown(): Promise<void> {
    const agents = Array.from(this.activeSessions.keys());
    await Promise.all(agents.map((id) => this.cleanup(id)));
  }

  /**
   * Determine the opencode agent mode based on the role.
   *
   * - ROUTA (planner): 'plan' mode - read-only, no file editing
   * - CRAFTER (implementer): 'build' mode - full tools enabled
   * - GATE (reviewer): 'plan' mode - read-only verification
   */
  private getModeForRole(role: AgentRole): string {
    switch (role) {
      case AgentRole.ROUTA:
        return "plan"; // Read-only planning mode
      case AgentRole.CRAFTER:
        return "build"; // Full implementation mode
      case AgentRole.GATE:
        return "plan"; // Read-only verification mode
      default:
        return "build"; // Default to build mode
    }
  }
}

